File: examples/graphql/src/test/java/example/armeria/server/graphql/GraphQLServiceTest.java
package example.armeria.server.graphql;

import static example.armeria.server.graphql.Main.newServer;
import static net.javacrumbs.jsonunit.fluent.JsonFluentAssert.assertThatJson;
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import com.google.common.base.Charsets;

import com.linecorp.armeria.client.WebClient;
import com.linecorp.armeria.common.AggregatedHttpResponse;
import com.linecorp.armeria.common.HttpMethod;
import com.linecorp.armeria.common.HttpStatus;
import com.linecorp.armeria.common.MediaType;
import com.linecorp.armeria.common.RequestHeaders;
import com.linecorp.armeria.server.Server;

class GraphQLServiceTest {

    private static Server server;
    private static WebClient client;

    @BeforeAll
    static void beforeClass() {
        server = newServer(0);
        server.start().join();
        client = WebClient.of("http://127.0.0.1:" + server.activeLocalPort());
    }

    @AfterAll
    static void afterClass() {
        if (server != null) {
            server.stop().join();
        }
        if (client != null) {
            client.options().factory().close();
        }
    }

    @ParameterizedTest
    @CsvSource({
            "{user(id: 1) {name}},hero",
            "{user(id: 2) {name}},human",
            "{user(id: 3) {name}},droid"
    })
    void testUserDataFetch(String query, String expected) throws Exception {
        final RequestHeaders headers = RequestHeaders.builder()
                                                     .path("/graphql")
                                                     .method(HttpMethod.POST)
                                                     .contentType(MediaType.GRAPHQL)
                                                     .build();
        final AggregatedHttpResponse response = client.execute(headers, query, Charsets.UTF_8)
                                                      .aggregate().get();

        assertThat(response.status()).isEqualTo(HttpStatus.OK);
        assertThatJson(response.contentUtf8()).node("data.user.name").isEqualTo(expected);
    }
}

File: examples/graphql-kotlin/src/test/kotlin/example/armeria/server/graphql/kotlin/GraphqlServiceTest.kt
package example.armeria.server.graphql.kotlin

import com.expediagroup.graphql.client.jackson.GraphQLClientJacksonSerializer
import com.expediagroup.graphql.client.types.GraphQLClientRequest
import com.linecorp.armeria.server.ServerBuilder
import com.linecorp.armeria.testing.junit5.server.ServerExtension
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.extension.RegisterExtension
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import kotlin.reflect.KClass

class GraphqlServiceTest {

    companion object {

        @JvmField
        @RegisterExtension
        val server: ServerExtension = object : ServerExtension() {
            @Throws(Exception::class)
            override fun configure(sb: ServerBuilder) {
                configureService(sb)
            }
        }

        private fun client(): GraphqlArmeriaClient {
            return GraphqlArmeriaClient(
                    uri = server.httpUri().resolve("/graphql"),
                    serializer = GraphQLClientJacksonSerializer())
        }
    }

    @ParameterizedTest
    @CsvSource(
            "1,hero",
            "2,human",
            "3,droid"
    )
    fun testUserDataFetch(id: String, expected: String) {
        runBlocking {
            val response = client().execute(UserNameQuery(id))
            assertThat(response.data?.userById?.name).isEqualTo(expected)
        }
    }

    data class UserNameResult(val userById: User)

    class UserNameQuery(id: String) : GraphQLClientRequest<UserNameResult> {
        override val query: String = "query {userById(id: $id) {name}}"
        override val operationName: String? = null
        override fun responseType(): KClass<UserNameResult> = UserNameResult::class
    }
}
